arch: arm: mpu: add configurable cache policies for armv7m and v8

lucien-nxp opened the pull request:
Add Kconfig options for FLASH and SRAM cache policy selection:

- CONFIG_ARM_MPU_SRAM_WRITE_THROUGH: enables Write-Through cache policy for SRAM regions instead of default Write-Back

Includes corresponding MPU attribute macros for ARMv7-M and ARMv8-M architectures. Maintains backward compatibility with existing configurations.

Validation performed on NXP RT1180 platform on arm_mpu_wt case

lucien-nxp committed:
arch: arm: mpu: Add Kconfig options for SRAM Write-Through cache policy

- CONFIG_ARM_MPU_SRAM_WRITE_THROUGH: enables Write-Through cache policy
  for SRAM regions instead of default Write-Back

Includes corresponding MPU attribute macros for ARMv7-M and ARMv8-M
architectures. Maintains backward compatibility with existing
configurations.

Signed-off-by: Lucien Zhao <lucien.zhao@nxp.com>

JarmouniA commented on arch/arm/core/mpu/Kconfig:
```diff
@@ -79,6 +79,20 @@ config ARM_MPU_PXN
 
+config ARM_MPU_FLASH_WRITE_BACK
+	bool "Use Write-Back cache policy for FLASH regions"
+	default n
```
```suggestion
```

lucien-nxp replied:
Updated.

JarmouniA commented on arch/arm/core/mpu/Kconfig:
```diff
@@ -79,6 +79,20 @@ config ARM_MPU_PXN
+
+config ARM_MPU_SRAM_WRITE_THROUGH
+	bool "Use Write-Through cache policy for SRAM regions"
+	default n
```
```suggestion
```

lucien-nxp replied:
Updated.

JarmouniA commented on arch/arm/core/mpu/arm_mpu_regions.c:
```diff
@@ -9,27 +9,39 @@
 #include <zephyr/arch/arm/mpu/arm_mpu_mem_cfg.h>
 
+#ifdef CONFIG_ARM_MPU_FLASH_WRITE_BACK
+#define FLASH_REGION_ATTR REGION_FLASH_WB_ATTR
```
```suggestion
#define ARM_MPU_FLASH_REGION_ATTR REGION_FLASH_WB_ATTR
```

hakehuang commented:
> Current approach does not modify any default configurations. If users have specific requirements for cache policies, they can change them through overlay or hard coding methods to avoid the need for multiple regression tests to cover the supported cases.

I would suggest add one test for this, maybe mpu_test with WT for mimxrt1180

lucien-nxp committed:
tests: arch: arm: add ARM MPU write-through test

Add test suite for ARM MPU write-through cache functionality including:
- Cache coherency validation with dma ways
- Cache coherency validation

Support case on mimxrt1180_evk cm33/cm7 cores

Signed-off-by: Lucien Zhao <lucien.zhao@nxp.com>

hakehuang commented on tests/arch/arm/arm_mpu_wt/src/main.c:
```diff
@@ -0,0 +1,526 @@
+	/* Test multiple patterns in user mode */
+	for (int i = 0; i < TEST_PATTERN_COUNT; i++) {
+		*test_addr = test_patterns[i];
+		barrier_dsync_fence_full();
```
I do not thing it is good if we already have a cache system, and we need resync cache every time.

hakehuang commented on tests/arch/arm/arm_mpu_wt/boards/mimxrt1180_evk_mimxrt1189_cm33.overlay:
```diff
@@ -0,0 +1,11 @@
+/*
```
need demo how to use mem_attr to set the cache policy

lucien-nxp replied:
I have added CONFIG_ARM_MPU_SRAM_WRITE_THROUGH in prj.config default. Not all platforms will set the MPU attribute in DTS, and the attributes of the sram/flash nodes will be configured by the array in mpu_region.c based on the existing framework of Zephyr. Therefore, this case only tests the wt attribute of the sdram node.

hakehuang replied:
the problem is that, in many user case, we do not need to configure the whole sdram as WT, WB would be much prefer.

lucien-nxp replied:
Shouldn't the addition be just a simple test case? I‘m not sure we need to be concerned about the specific usage scenarios of the users. This case is only for the default verification of the zephyr, sram node configuration into WT mode and the related tests. Moreover, in my patch, I have already provided REGION_RAM_WT_ATTR/DT_MEM_ARM_MPU_RAM_WT, which can be configured by the users in the DT.

lucien-nxp replied:
I have added code to support DT configuration. CM33 and CM7 use different ways to configuration memory with WT attribution.

wearyzen commented on include/zephyr/dt-bindings/memory-attr/memory-attr-arm.h:
```diff
@@ -40,6 +41,7 @@
 #define DT_MEM_ARM_MPU_EXTMEM      DT_MEM_ARM(ATTR_MPU_EXTMEM)
 #define DT_MEM_ARM_MPU_RAM_PXN     DT_MEM_ARM(ATTR_MPU_RAM_PXN)
 #define DT_MEM_ARM_MPU_DEVICE      DT_MEM_ARM(ATTR_MPU_DEVICE)
+#define DT_MEM_ARM_MPU_RAM_WT      DT_MEM_ARM(ATTR_MPU_RAM_WT)
```
You are right.
The problem is that this layer has I think 12 bits for attributes and with current design it gives us 12 options, 8 of which would have been taken after this PR, so I was thinking of redesigning it as per the link I shared.
The naming suggestion was to avoid any breaking change for the macro users but I think I can figure that out later. I am ok with this name.

wearyzen reviewed (commented):
I think the test could be made simpler.
Important part is to check if memory attributes are set correctly and if WT works as expected (either by volatile without a sync or a dma) but I am not sure if subtests like userspace etc. are adding any value here.
Also, does this run in Zephyr ci since I don't see any qemu/fvp boards as allowed? if not then should we even have this test?

lucien-nxp commented on tests/arch/arm/arm_mpu_wt/boards/mimxrt1180_evk_mimxrt1189_cm7.config:
```diff
@@ -0,0 +1 @@
+CONFIG_ARM_MPU_SRAM_WRITE_THROUGH=y
```
Hake suggest we need also to provide a solution for some regions defined by DT, `CONFIG_ARM_MPU_SRAM_WRITE_THROUGH` only work for zephyr,sram node. So I set `CONFIG_ARM_MPU_SRAM_WRITE_THROUGH` in [mimxrt1180_evk_mimxrt1189_cm7.config](https://github.com/zephyrproject-rtos/zephyr/pull/99238/files/2b3b60eb3c28610e289b15ddfb825f46beabe72f#diff-e74fbd22e2bb0570c163e9097ffa658f99ff90fb9381ca460e823fa7f7cc1675), and show DT mpu settings for cm33 core

lucien-nxp commented:
> Important part is to check if memory attributes are set correctly and if WT works as expected (either by volatile without a sync or a dma) but I am not sure if subtests like userspace etc. are adding any value here.

1. I agree the userspace and stack tests don't add significant value for WT verification, now I only keep three test unit to test cache configuration.
2. I have added qemu_cortex_m7 in testcase.yaml.

hakehuang reviewed (dismissed):
board testing pass in v4.3.0-2109-gaffd08466091

lucien-nxp commented on tests/arch/arm/arm_mpu_wt/src/main.c:
```diff
@@ -0,0 +1,548 @@
+/**
+ * @brief Test basic memory region read/write functionality
+ */
+ZTEST(arm_mpu_wt, test_sram_basic_access)
```
Thank you, I have removed this test unit.

JarmouniA commented on tests/arch/arm/arm_mpu_wt/testcase.yaml:
```diff
@@ -0,0 +1,15 @@
+    arch_allow:
+      - arm
+    filter: CONFIG_ARM_MPU
+    platform_allow:
```
```suggestion
    integration_platforms:
```
